### ISSUE TITLE
model.save() fails when model has been populated

### ISSUE THREAD

**Issue by drbolsen** [balderdashy/waterline#975]
The issue is similar to an already reported issue in sails repo [2073](https://github.com/balderdashy/sails/issues/2073) - updating both a main model and an association in one go will return an error in callback and an empty result value however both models are actually updated. Weird.

The code:
```
    Passport.findOne().where({ protocol : 'local' })
      .populate('user', {where : {
                          resetPasswordToken: req.params.token, 
                          resetPasswordExpires : {'>=': new Date (Date.now())}
                          }
                        }
               ) // populate user associations for further use to change password
      .exec(function(err, passport){

        ...
        <skipped>
        ...

        passport.user.resetPasswordToken = undefined;
        passport.user.resetPasswordExpires = undefined;
        passport.password = req.body.password;
        passport.save(function (err, saved) {
          return next (err, saved);  // error contains error below, saved is undefined both models are actually updated
        });
```
The error:
```
rror (E_VALIDATION) :: 1 attribute is invalid
    at WLValidationError.WLError (/projects/csh-boilerplate-app/node_modules/sails/node_modules/waterline/lib/waterline/error/WLError.js:26:15)
    at new WLValidationError (/projects/csh-boilerplate-app/node_modules/sails/node_modules/waterline/lib/waterline/error/WLValidationError.js:20:28)
    at /projects/csh-boilerplate-app/node_modules/sails/node_modules/waterline/lib/waterline/query/validate.js:45:43
    at allValidationsChecked (/projects/csh-boilerplate-app/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:203:5)
    at done (/projects/csh-boilerplate-app/node_modules/sails/node_modules/async/lib/async.js:135:19)
    at /projects/csh-boilerplate-app/node_modules/sails/node_modules/async/lib/async.js:32:16
    at /projects/csh-boilerplate-app/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:194:14
    at done (/projects/csh-boilerplate-app/node_modules/sails/node_modules/async/lib/async.js:135:19)
    at /projects/csh-boilerplate-app/node_modules/sails/node_modules/async/lib/async.js:32:16
    at /projects/csh-boilerplate-app/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:157:64
    at /projects/csh-boilerplate-app/node_modules/sails/node_modules/async/lib/async.js:125:13
    at Array.forEach (native)
    at _each (/projects/csh-boilerplate-app/node_modules/sails/node_modules/async/lib/async.js:46:24)
    at Object.async.each (/projects/csh-boilerplate-app/node_modules/sails/node_modules/async/lib/async.js:124:9)
    at validate (/projects/csh-boilerplate-app/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:156:11)
    at /projects/csh-boilerplate-app/node_modules/sails/node_modules/async/lib/async.js:125:13

Invalid attributes sent to Passport:
 • user
   • "required" validation rule failed for input: null
```

**Comment by dmarcelino** [comment]
Hi @drbolsen, this seems the same problem as #964, please read @devinivy's response https://github.com/balderdashy/waterline/issues/964#issuecomment-95626676 and let us know if that addresses your issue.

**Comment by drbolsen** [comment]
Thanks for a prompt response mate, much appreciated. I will look at it more carefully tomorrow, however I have experimented a bit and found that by adding the following line to my code solves the problem:
```
passport.password = req.body.password;
passport.user = passport.user.id;  <---- This fixes the issue in my case
passport.save(function (err, saved) {

```

I thought it wouldn't be necessary as this was already part of the populated collection

```json
{"user":{
"username":"Tester",
"email":"Tester@test.com",
"resetPasswordToken":"76bb160e-cb57-40ac-829c-9d42584ea3f8",
"resetPasswordExpires":"2015-04-25T12:28:28.154Z",
"title":"Mr",
"id":1,
"createdAt":"2015-04-25T10:15:09.795Z",
"updatedAt":"2015-04-25T11:28:28.218Z"
},
"protocol":"local",
"password":"$2a$10$8AdCKFwAUKR0xGElMUFsLefNHn4BtvSu1Ch/LVFLxIx.j7iQqE9O6",
"accessToken":"LRNfpDFTzwTT3WEEpx2K2dPxbo2ePgbmtIkn8OcKgi/R9jrqWqA2I0Gf4ywvSAV8",
"id":1,
"createdAt":"2015-04-25T10:15:10.077Z",
"updatedAt":"2015-04-25T11:24:48.398Z"}
```

but it appears that an association attribute that links two models must be specified.

**Comment by drbolsen** [comment]
Following up on this one, further digging into the issue I found that my fix is not actually a fix - it just mitigates an apparent exception error message. In reality assigning "id", which is in essence a foreign key that links two models, will just re-link the populated association to the updated record. No changes are made. So in my particular scenario - resetPasswordToken will never be updated.  But it seems like an expected behaviour though. And I have no issues with this.

My original issue - an exception error  __"required" validation rule failed for input: null"__ is a bit different, I looked into sources and found that it had been caused by this piece of code:

sails/node_modules/waterline/lib/waterline/utils/nestedOperations/update.js

```javascript

  438      // Run an update operation to set the foreign key to null on all the
  439      // associated child records.
  440:     values[operation.keyName] = null; <---- This will set an attribute to null
  441      
  442      self.waterline.collections[operation.model].update(operation.criteria, values).exec(next);

```

Basically if a key set to __required__ like in my model:

User.js

```javascript
var User = {
  // Enforce model schema in the case of schemaless databases
  schema: true,

  attributes: {
    username  : { type: 'string', unique: true },
    email     : { type: 'email',  unique: true },
    passports : { collection: 'Passport', via: 'user' }
  }
};
```

Passport.js

```javascript
var Passport = {
  attributes: {

    protocol: { type: 'alphanumeric', required: true },

    password    : { type: 'string', minLength: 8 },
    accessToken : { type: 'string' },


    provider   : { type: 'alphanumericdashed' },
    identifier : { type: 'string' },
    tokens     : { type: 'json' },

    user: { model: 'User', required: true },  <-------- Here 

    validatePassword: function (password, next) {
      bcrypt.compare(password, this.password, next);
    }
  },
```

It will trigger an exception because the update happens in few rounds - first round updates all associated child records, the next unlink them by setting foreign keys to __null__ and then reconnect them again:

```javascript
function sync(parents, operations, cb) {
  var self = this;

  async.auto({

    // Update any nested associations
    update: function(next) {
      updateRunner.call(self, parents, operations, next);
    },

    // For each parent, unlink all the associations currently set
    unlink: ['update', function(next) {
      unlinkRunner.call(self, parents, operations, next);
    }],

    // For each parent found, link any associations passed in by either creating
    // the new record or linking an existing record
    link: ['unlink', function(next) {
      linkRunner.call(self, parents, operations, next);
    }]

  }, cb);
}

```

That is why I had my models updated regardless of exception.

**Comment by devinivy** [comment]
That's right!  I'd love a solid proposed solution to this issue.  The only solution currently known is to minimize unlinking/linking routines by performing an additional `find` then diffing nested records.  Bearing in mind that `Model#save` eventually defers to `Collection#update`, I would love some effort to go into figuring out if there's duplicate work being done here (https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/dql/update.js#L47) as in the nested update routine mentioned above (https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/dql/update.js#L215).

**Comment by dmarcelino** [comment]
I don't have a solid proposal, but I have been thinking about this last night and maybe there is a way to avoid that additional `find`... @devinivy, in the other thread you said:
> 1. When an id is specified on a nested record passed to `Collection#update`, update those nested records.
> 2. Remove all associations with the parent record, either by removing associations from a junction table (m:n) or by setting child record foreign keys to null (1:n).
> 3. For those nested records passed to `Collection#update` without specified without ids, create them.
> 4. Gather ids from all nested records, now created or updated, and associate those records with the parent record.

What if we instead we did:
1. When an id is specified on a nested record passed to `Collection#update`, update those nested records.
2. Remove **associations with the parent record which are not in the values supplied in the `.update()` query**, either by removing associations from a junction table (m:n) or by setting child record foreign keys to null (1:n). **In other words remove only the associations that need to be removed.**
3. For those nested records passed to `Collection#update` without specified without ids, create them.
4. Gather ids from all nested records, now created or updated, and associate those records with the parent record. **The problem now is that the previous existing associations are still there, while we need to add new associations and don't know which ones are new**.

Another option is to wait until transactions (#755) are out and do that extra `.find()` in a transaction with some/all of this operations...

**Comment by dmarcelino** [comment]
See also https://github.com/balderdashy/waterline/issues/751#issue-50808660, judging from the `beforeValidate` calls nested updates seem to be behaving awkwardly...

**Comment by sailsbot** [comment]
Thanks for posting, @drbolsen.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by albertpeiro** [comment]
I'm finding this issue bothering. From the point of view of the update method, I wouldn't mind it to not accept attributes with populated models. It would force us all to send only ids.

But since it does accept populated associations, then the expected behaviour would be for them to just work, regardless of the attribute being "required" or not. It just looks very inconsistent and buggy as it is.

How can we fix it? Are there any plans for Sails 1.0?

**Comment by ejames17** [comment]
@albertpeiro  has anyone fixed this issue? Seems to be a quite a bother
